File: 5373.py
class Solution:
    def findMinFibonacciNumbers(self, k: int) -> int:
        arr = [1,1]
        f1 = 1
        f2 = 1
        while True:
            f = f1+f2
            if f > k:
                break
            arr.append(f)
            f1 = f2
            f2 = f
        ret = 0
        stop = len(arr)-1
        while True:
            for i in range(stop,-1,-1):
                if arr[i] <= k:
                    ret += 1
                    k -= arr[i]
                    stop = i
                    break
            if k == 0:
                break

        return ret


s = Solution()

print(s.findMinFibonacciNumbers(7))
print(s.findMinFibonacciNumbers(10))
print(s.findMinFibonacciNumbers(19))
print(s.findMinFibonacciNumbers(1000000000))

            

File: 5373.c
#include<stdio.h>

int find(int l,int h,int *arr,int x){
    int mid = (h+l)/2;
    if(l >= h){
        return x - arr[mid];
    }        
    if(x < arr[mid]){
        return find(l,mid,arr,x);
    }
                
    if (x > arr[mid]){
        return find(mid,h,arr,x);
    }
    return 0;
}

int findMinFibonacciNumbers(int k){
    int arr[50];
    int index = 2;
    int l,h,ret;
    arr[0] = 1;
    arr[1] = 1;
    while(1){
        int f = arr[index-1] + arr[index-2];
        arr[index] = f;
        index ++;
        if(f >= k){
            break;
        }
    }
    printf("index %d\n",index);

    l = 0;
    h = index-1;
    ret = 0;
    while(1){
        int last = find(l,h,arr,k);
        
        printf("%d----%d \n",last,k);
        
        ret += 1;
        if(last == 0){
            break;
        }
        k = last;
    }
    return ret; 
}

int main(){
    printf("%d\n",findMinFibonacciNumbers(100000000));
}